[Nicholas Lam] iP

niclammm opened the pull request:
# Repsmax - Task Tracking Tool

## Description

Repsmax is a task tracking tool that is easy to use. It provides a simple and efficient way to manage your tasks and stay organized.  `Download it` [here!](https://github.com/niclammm/ip) 

### Features

- Easy to use
- Easy to learn
- ~~Fast~~ Super fast to learn

### Types of Events It Can Handle

1. **Todo**
2. **Deadline**
3. **Event**

### Example Usage

```java
// Example of how to create tasks in Repsmax
TaskList tasks = new TaskList();
tasks.add(new Todo("Read book"));
tasks.add(new Deadline("Submit report", "2024-09-01"));
tasks.add(new Event("Team meeting", "2024-08-30T10:00", "2024-08-30T11:00"));
```

###Tasks to be Done`🚀`
- [ ] Add user interface enhancements
- [ ] Improve task handling
- [ ] Update documentation
>These will be done soon! 

### Latest Features 
- [x] Added better exception handling abilities

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

niclammm committed:
Removed Duke and added Repsmax

niclammm committed:
Added Scanning function and changed system outputs

niclammm committed:
Added String array to hold inputs by scanner

niclammm committed:
added new class Task

niclammm committed:
Adjusted main to use new object Task as well as added the mark and unmark function

niclammm committed:
Added new subclasses of Task

niclammm committed:
Added usage of new classes

niclammm committed:
Added Repsmax exception class

niclammm committed:
Integrated try catch blocks for array out of bound and empty statements

niclammm committed:
added delete

niclammm committed:
Added Automated Text UI Testing

niclammm committed:
Added file output and input functions in Event Classes

niclammm committed:
Added file loading and reading functions

niclammm committed:
Added data.txt file

niclammm committed:
no message

niclammm committed:
Merge branch 'branch-Level-7'

niclammm committed:
Added deadline function and parsing

niclammm committed:
Merge branch 'branch-Level-8'

niclammm committed:
Add Parser Storage TaskList Ui class

niclammm committed:
Merge branch 'A-MoreOOP'

niclammm committed:
Updated changes to previous classes

niclammm committed:
Bug fixes due to intellij

niclammm committed:
Add all classes into the repsmax package

niclammm committed:
Merge branch 'add-gradle-support'

niclammm committed:
Add A-Gradle

niclammm committed:
Add Javadoc

niclammm committed:
Add todo, storage and task test

niclammm committed:
Add find function

niclammm committed:
Merge branch 'Level-9'

niclammm committed:
Change java documentation

niclammm committed:
Merge branch 'branch-A-JavaDoc'

niclammm committed:
Commit for branch

Anvita2110 commented on src/main/java/repsmax/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+public class Deadline extends Task {
+
+    /** The date/time by which the task should be completed. */
+    protected String by;
```
Any particular reason for using **protected** and not **private**?
```suggestion
    protected String by; Relevance of access modifier
```

Anvita2110 commented on src/main/java/repsmax/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    @Override
+    public String toFileFormat() {
+        return "D | " + (isDone() ? "1" : "0") + " | " + getDescription() + " | " + by;
```
Do you think String.format( ) better way than concatenating?

Anvita2110 commented on src/main/java/repsmax/Repsmax.java:
```diff
@@ -0,0 +1,61 @@
+package repsmax;
```
```suggestion
package repsmax; Neat coding style
```
Great job on writing such a neat code!
Really Impressed by the clarity enabling efficient code reading. 🎉

Anvita2110 commented on src/main/java/repsmax/Repsmax.java:
```diff
@@ -0,0 +1,61 @@
+        tasks = new TaskList();
+        try {
+            storage.load(tasks);
+        } catch (Exception e) {
```
Isn't it better to handle more specific ( IOException or RepsmaxException ) error than capturing general Exception?

Anvita2110 commented on src/main/java/repsmax/Repsmax.java:
```diff
@@ -0,0 +1,61 @@
+     * @param args Command line arguments (not used).
+     */
+    public static void main(String[] args) {
+        new Repsmax("C:/Users/nicla/OneDrive/Desktop/Cs2103/repo/src/data.txt").run();
```
What could be better here: 
* absolute path
* relative path

Anvita2110 commented on src/main/java/repsmax/Event.java:
```diff
@@ -0,0 +1,47 @@
+    public String toFileFormat() {
+        return "E | " + (super.isDone() ? "1" : "0") + " | " + getDescription() + " | " + from + " | " + to;
+    }
+}
```
Good job here, Don't see any coding style violations

Anvita2110 commented on src/main/java/repsmax/Parser.java:
```diff
@@ -0,0 +1,258 @@
+     * @param storage   The storage handler for saving and loading tasks.
+     */
+    public void parse(String userInput, TaskList tasks, Ui ui, Storage storage) {
+        String[] splitInput = userInput.split(" ", 2);
```
splitInput sounds like a verb, I suggest if you might want to change it to a plural noun word, like Inputs since it is a collection of items.

Anvita2110 commented on src/main/java/repsmax/Parser.java:
```diff
@@ -0,0 +1,258 @@
+    private void handleEventCommand(String[] splitInput, TaskList tasks, Ui ui, Storage storage) {
+        try {
+            String[] parts = splitInput[1].split("/from ", 2);
+            String[] fromTo = parts[1].split("/to ", 2);
```
>Just a suggestion: 
could use variable name "timing" as fromTo might not sound meaningful to a novice user

Anvita2110 commented on src/main/java/repsmax/Parser.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+}
+
```
Good work on coding style with minute suggestions only! 😊

Anvita2110 commented on src/main/java/repsmax/Storage.java:
```diff
@@ -0,0 +1,79 @@
+            System.out.println("An error occurred while loading tasks.");
+        }
+    }
+}
```
All cases well handles with good exception handling

> [!TIP]
> Is it ok to handle exceptions in few places with just printing a message or would it better on using printing exception.getMessage() or exception.toString()?
As shown below:

![image for showing exception.getMessage() or exception.toString()](https://github.com/user-attachments/assets/a586a23a-ee87-499d-b570-c06b0d6a4f66)

Anvita2110 commented on src/main/java/repsmax/Task.java:
```diff
@@ -0,0 +1,122 @@
+ * the task to and from a file format.
+ * </p>
+ */
+public class Task {
```
Any reason why you chose to make Task a concrete class and not abstract?

Anvita2110 commented on src/main/java/repsmax/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.List;
+import java.util.ArrayList;
+import repsmax.Task;
+import java.util.stream.Collectors;
```
According to java coding standards, I think this has to come before repsmax.Task
Kindly reconfirm

Anvita2110 commented on src/main/java/repsmax/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+        return tasks.stream()
+                .filter(task -> task.getDescription().toLowerCase().contains(keyword.toLowerCase()))
+                .collect(Collectors.toList());
+    }
```
> [!IMPORTANT]
> Were we supposed to implement find feature in iP right now?

Anvita2110 commented on src/main/java/repsmax/Ui.java:
```diff
@@ -0,0 +1,113 @@
+package repsmax;
+
+import java.util.Scanner;
+import java.util.List;
```
Please note the import order and change it accordingly

Anvita2110 reviewed (commented):
## Great job on coding style

* Neatly written code
* Correctly modularised
* It is easy to read
* Good exception handling

Only a few minor standard coding violations found and a few suggestions given!
Also put few questions regarding why you chose this way of coding for better understanding on those topics, hope it will help us both to gain knowledge about it!

gohsl99 commented on src/main/java/repsmax/Parser.java:
```diff
@@ -0,0 +1,258 @@
+     */
+    private void handleDeadlineCommand(String[] splitInput, TaskList tasks, Ui ui, Storage storage) {
+        try {
+            String[] parts = splitInput[1].split("/by ", 2);
```
Perhaps commandParts or inputSegments can be a better choice for parts here.

gohsl99 commented on src/main/java/repsmax/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return the number of tasks in the list.
+     */
+    public int size() {
```
Perhaps getSize() is a more consistent name here.

gohsl99 commented on src/main/java/repsmax/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     * @return the {@code Task} that was removed.
+     * @throws IndexOutOfBoundsException if the index is out of range.
+     */
+    public Task remove(int index) {
```
Perhaps removeTask() or removeTaskAtIndex() is a more consistent name here.

gohsl99 commented on src/main/java/repsmax/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     * @param keyword the keyword to search for in task descriptions.
+     * @return a list of tasks that contain the keyword in their description.
+     */
+    public List<Task> find(String keyword) {
```
Perhaps findTask() or findTaskWithKeyword() is a more consistent name here.

gohsl99 commented on src/test/java/repsmax/StorageTest.java:
```diff
@@ -0,0 +1,47 @@
+public class StorageTest {
+
+    @Test
+    public void testSave_andLoad_tasksArePersisted() throws Exception {
```
Perhaps a clearer description of test inputs can be used for method name (unitBeingTested_descriptionOfTestInputs_expectedOutcome).

gohsl99 commented on src/test/java/repsmax/TodoTest.java:
```diff
@@ -0,0 +1,20 @@
+package repsmax;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps just import the specific class instead of using *.

gohsl99 commented on src/test/java/repsmax/TaskTest.java:
```diff
@@ -0,0 +1,33 @@
+package repsmax;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps just import the specific class instead of using *.

gohsl99 commented on src/test/java/repsmax/StorageTest.java:
```diff
@@ -0,0 +1,47 @@
+import java.io.File;
+import java.io.FileWriter;
+import java.util.Scanner;
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps just import the specific class instead of using *.

gohsl99 commented on src/main/java/repsmax/Task.java:
```diff
@@ -0,0 +1,122 @@
+     * @return a {@code Task} object created from the file format string.
+     * @throws IllegalArgumentException if the file format string is invalid.
+     */
+    public static Task fromFileFormat(String fileFormat) {
```
Perhaps the method name should include verb to make it more understandable.

gohsl99 reviewed (commented):
Overall LGTM, most of the method and variable names are good to understand and has good code quality for naming. Keep up the good work!

niclammm committed:
Add Gui functionality

niclammm committed:
Merge branch 'branch-A-CodingStandard'

niclammm committed:
Change documentation

niclammm committed:
Add assert feature

Assertions help catch bugs during the development phase by checking assumptions that the code relies on. It also makes the code easier to understand and maintain.
Reasons for adding assertion in the Repsmax, Storage, Task and TaskList Class:
Repsmax: helps to identify empty file path and user input
Storage: helps to identify empty file path, failure of file path and existence of the file
Task: helps to identify empty descriptions
TaskList: ensures that task input is not a null

niclammm committed:
Improve code quality for Parser, RepsMax, Storage, Task, MainWindow classes

MainWindow Class
Add basic error handling in handleUserInput.
Add comments for better clarity.
Include a null check for repsmax.

Parser Class
Update comments to use <p> tags for better readability and added descriptions for parameters and return values.
Add spaces after periods for better readability in output messages.
Adopt consistent error messages across methods.

RepsMax Class
Class-Level Documentation: Update comments for GUI-functionalities added previously

Storage Class
Error Messages: Include exception messages to provide more specific error details.
Add and refined comments for clarity, explaining the purpose of each method and key operations.
Clean up messy try catch blocks

Task Class
Add the invalid format string to the exception message for better debugging.
Ensure comments are clear and consistent

niclammm committed:
Add stream usage and explanation

niclammm committed:
Merge pull request #2 from niclammm/branch-A-CodeQuality

Branch a code quality

niclammm committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/repsmax/Repsmax.java
#	src/main/java/repsmax/Storage.java

niclammm committed:
Add new priority input and reading to the Event class and all its subclasses

niclammm committed:
Add new parameter input for the new priority functionality

niclammm committed:
Testing something

niclammm committed:
Testing something

niclammm committed:
Organise the code for better readability Fix the bugs related to the event and deadline classes

niclammm committed:
Improve coding quality and Javadoc explanations

niclammm committed:
Improve coding quality and Javadoc descriptions

niclammm committed:
Shift date parsing logic to the Event and Deadline classes for better readability and shorter code

niclammm committed:
Change README descriptions and Add Ui.png

niclammm committed:
Make file path general so others can use the JAR file and add more exceptions handling for the delete and mark functions

niclammm committed:
Update README file user guide descriptions

niclammm committed:
Change font sizes for description

niclammm committed:
Add picture into User Guide

niclammm committed:
Change link to image to the correct link

niclammm committed:
Alter image size

niclammm committed:
Correct link mistake

niclammm committed:
Improve exception handling
